unittests/replay_block_invariants_tests.cpp

```diff
@@ -0,0 +1,252 @@
+//      with --force-all-checks
+
+BOOST_AUTO_TEST_SUITE(replay_block_invariants_tests)
+
+using namespace eosio::testing;
+using namespace eosio::chain;
+using namespace fc::crypto;
+
+struct test_fixture {
+   eosio::testing::tester chain;
+
+   // Creates blocks log
+   test_fixture() {
+      // Create a few accounts and produce a few blocks to fill in blocks log
+      chain.create_account("replay1"_n);
+      chain.produce_blocks(1);
+      chain.create_account("replay2"_n);
+      chain.produce_blocks(1);
+      chain.create_account("replay3"_n);
+
+      chain.produce_blocks(10);
+
+      // Make sure the accounts were created
+      BOOST_REQUIRE_NO_THROW(chain.get_account("replay1"_n));
+      BOOST_REQUIRE_NO_THROW(chain.get_account("replay2"_n));
+      BOOST_REQUIRE_NO_THROW(chain.get_account("replay3"_n));
+
+      // Stop the node and save blocks_log
+      chain.close();
+   }
+
+   // Removes state directory
+   void remove_existing_states(std::filesystem::path& state_path) {
+      std::filesystem::remove_all(state_path);
+      std::filesystem::create_directories(state_path);
+   }
+
+   // Corrupts the signature of last block which attaches a QC in the blocks log
+   void corrupt_qc_signature_in_block_log() {
+      controller::config config      = chain.get_config();
+      auto               blocks_dir  = chain.get_config().blocks_dir;
+      auto               qc_ext_id   = quorum_certificate_extension::extension_id();
+
+      block_log blog(blocks_dir, chain.get_config().blog);
+
+      // find the first block which has QC extension starting from the end of
+      // block log and going backward
+      uint32_t block_num = blog.head()->block_num();
+      while (!blog.read_block_by_num(block_num)->contains_extension(qc_ext_id)) {
+         --block_num;
+         BOOST_REQUIRE(block_num != 0);
+      }
+
+      // clone the QC block
+      auto qc_block = std::make_shared<signed_block>(blog.read_block_by_num(block_num)->clone());
+      BOOST_TEST(qc_block);
+
+      // remove qc block until the end from block log
+      BOOST_REQUIRE_NO_THROW(block_log::trim_blocklog_end(blocks_dir, block_num-1));
+      BOOST_REQUIRE_NO_THROW(block_log::smoke_test(blocks_dir, 1));
+
+      // retrieve QC block extension from QC block
+      auto block_exts = qc_block->validate_and_extract_extensions();
+      auto qc_ext_itr = block_exts.find(qc_ext_id);
+      auto& qc_ext    = std::get<quorum_certificate_extension>(qc_ext_itr->second);
+      auto& qc        = qc_ext.qc;
+
+      // remove QC block extension from QC block
+      auto& exts = qc_block->block_extensions;
+      std::pair<uint16_t,vector<char>> target{qc_ext_id, {}};
+      auto itr = std::lower_bound(exts.begin(), exts.end(), target, [](const auto& ext1, const auto& ext2){
+         return ext1.first < ext2.first;
+      });
+      BOOST_REQUIRE(itr != exts.end());
+      qc_block->block_extensions.erase(itr);
```

Comment:
I think instead of this lookup just use [std::erase_if](https://en.cppreference.com/w/cpp/container/vector/erase2)